WF check lifetime bounds for locals with type params

ShoyuVanilla opened the pull request:
FCP proposal in https://github.com/rust-lang/rust/pull/149389#issuecomment-3669689488

Fixes rust-lang/rust#115175
Fixes rust-lang/rust#148854

ShoyuVanilla committed:
WF check lifetime bounds for locals with type params

ShoyuVanilla commented:
@craterbot check

petrochenkov commented:
@craterbot p=11
All the small runs will complete in a few hours and won't need to wait for a few days until the long beta runs complete.

lcnr commented on tests/ui/regions/wf-bound-region-in-local-issue-115175.rs:
```diff
@@ -0,0 +1,11 @@
+fn test<T>() {
+    let _ = None::<Static<T>>;
+    //~^ ERROR the parameter type `T` may not live long enough
+}
```
can you also add a test for the unsoundness?

lcnr commented on tests/ui/regions/wf-bound-region-in-local-soundness-issue-148854.rs:
```suggestion
        // This has to error due to an unsatisfied outlives bound on
        // `Inner<T: 'static>` as its implicit drop relies on that
        // bound.
```

lcnr reviewed (commented):
writing an FCP for this rn

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/149389/3f4dc1e02d759aa3c3807d4efc1f7f6e293536a5..223927a17758d9507320521d95400e052e952f0e)*

rust-timer commented:
Finished benchmarking commit (f70ee28831acc074c109afab32d287057c3eca46): [comparison URL](https://perf.rust-lang.org/compare.html?start=cf8a95590a1b768b7711f2631d5b5e3ead464de7&end=f70ee28831acc074c109afab32d287057c3eca46&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - please read the text below


Benchmarking this pull request means it may be perf-sensitive – we'll automatically label it not fit for rolling up. You can override this, but we strongly advise not to, due to possible changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please do so in sufficient writing along with `@rustbot label: +perf-regression-triaged`. If not, please fix the regressions and do another perf run. If its results are neutral or positive, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=cf8a95590a1b768b7711f2631d5b5e3ead464de7&end=f70ee28831acc074c109afab32d287057c3eca46&stat=instructions:u)
Our most reliable metric. Used to determine the overall result above. However, even this metric can be noisy.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.6%  | [0.2%, 1.0%]   | 2     |
| Regressions ❌ <br /> (secondary)  | 0.1%  | [0.1%, 0.1%]   | 1     |
| Improvements ✅ <br /> (primary)   | -0.4% | [-0.4%, -0.4%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.3%  | [-0.4%, 1.0%]  | 3     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=cf8a95590a1b768b7711f2631d5b5e3ead464de7&end=f70ee28831acc074c109afab32d287057c3eca46&stat=max-rss)
<details>
<summary>Results (primary -5.4%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -5.4% | [-5.4%, -5.4%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -5.4% | [-5.4%, -5.4%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=cf8a95590a1b768b7711f2631d5b5e3ead464de7&end=f70ee28831acc074c109afab32d287057c3eca46&stat=cycles:u)
<details>
<summary>Results (primary -1.3%, secondary -1.2%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.2%  | [2.2%, 2.2%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 2.3%  | [2.3%, 2.3%]   | 1     |
| Improvements ✅ <br /> (primary)   | -2.5% | [-3.0%, -2.2%] | 3     |
| Improvements ✅ <br /> (secondary) | -3.0% | [-3.8%, -2.2%] | 2     |
| All ❌✅ (primary)                 | -1.3% | [-3.0%, 2.2%]  | 4     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=cf8a95590a1b768b7711f2631d5b5e3ead464de7&end=f70ee28831acc074c109afab32d287057c3eca46&stat=size:linked_artifact)
<details>
<summary>Results (primary -0.1%, secondary -0.1%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.0%  | [0.0%, 0.0%]   | 6     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.2% | [-1.0%, -0.1%] | 5     |
| Improvements ✅ <br /> (secondary) | -0.1% | [-0.1%, -0.1%] | 32    |
| All ❌✅ (primary)                 | -0.1% | [-1.0%, 0.0%]  | 11    |
</details>

**Bootstrap**: 470.335s -> 475.664s (1.13%)
**Artifact size**: 386.93 MiB -> 386.89 MiB (-0.01%)

lqd commented:
Is it expected that changing WF has diffs in mir-opts tests?

lcnr commented:
> Is it expected that changing WF has diffs in mir-opts tests?

doesn't seem too surprising to me, even if I can't confidently say why you get the changes from this PR

whether locals have user type annotations should change the resulting MIR

ShoyuVanilla commented:
I looked into those mir-opt changes and the following is what I found out 😄 
In common, those differences are made while building THIR or MIR rather than during MIR passes, but visible in later passes in `tests/mir-opt/` because we only have output files for those passes in our test suite.

### Extra locals

We have extra locals like in the following test:

```diff
# tests/mir-opt/instsimplify/casts.redundant.InstSimplify-after-simplifycfg.diff
      let mut _4: *const &u8;
      scope 1 (inlined generic_cast::<&u8, &u8>) {
          let mut _5: *const &u8;
+         let mut _6: *const &u8;
      }
```

This originates from the extra THIR expression when lowering HIR's cast expr to thir expr in the following lines:

https://github.com/rust-lang/rust/blob/0d162b25edd5bf0dba9a22e83b614f1113e90474/compiler/rustc_mir_build/src/thir/cx/expr.rs#L1048-L1079

Here's the `fn generic_cast`:

```rust
fn generic_cast<T, U>(x: *const T) -> *const U {
    x as *const U
}
```

since `x: *const T` has a type param, this PR considers it as having user ty, pushes extra `thir::ExprKind::ValueTypeAscription` and it results in an extra local in MIR.

Here's what we get with `rustc -Zmir-opt-level=0 -Zunpretty=mir`

```diff
diff --git a/without_pr b/with_pr
index 2cf0332f086..fa94c259aaf 100644
--- a/without_pr
+++ b/with_pr
@@ -1,41 +1,49 @@
 fn generic_cast(_1: *const T) -> *const U {
     debug x => _1;
     let mut _0: *const U;
-    let mut _2: *const T;
+    let mut _2: *const U;
+    let mut _3: *const T;
 
     bb0: {
         StorageLive(_2);
-        _2 = copy _1;
-        _0 = move _2 as *const U (PtrToPtr);
+        StorageLive(_3);
+        _3 = copy _1;
+        _2 = move _3 as *const U (PtrToPtr);
+        _0 = copy _2;
+        StorageDead(_3);
         StorageDead(_2);
         return;
     }
 }
 
 fn redundant(_1: *const &u16) -> *const &u8 {
     debug x => _1;
     let mut _0: *const &u8;
     let mut _2: *const &u8;
     let mut _3: *const &u8;
     let mut _4: *const &u16;
     scope 1 (inlined generic_cast::<&u16, &u8>) {
-        let mut _5: *const &u16;
+        let mut _5: *const &u8;
+        let mut _6: *const &u16;
     }
 
     bb0: {
         StorageLive(_2);
         StorageLive(_3);
         StorageLive(_4);
         _4 = copy _1;
         StorageLive(_5);
-        _5 = copy _4;
-        _3 = move _5 as *const &u8 (PtrToPtr);
+        StorageLive(_6);
+        _6 = copy _4;
+        _5 = move _6 as *const &u8 (PtrToPtr);
+        _3 = copy _5;
+        StorageDead(_6);
         StorageDead(_5);
         StorageDead(_4);
         _2 = move _3 as *const &u8 (PtrToPtr);
         _0 = copy _2;
         StorageDead(_3);
         StorageDead(_2);
         return;
     }
 }
```

But those locals are completely gone with `-Zmir-opt-level=1`, regardless of the changes in this PR:

```
fn generic_cast(_1: *const T) -> *const U {
    debug x => _1;
    let mut _0: *const U;

    bb0: {
        _0 = copy _1 as *const U (PtrToPtr);
        return;
    }
}

fn redundant(_1: *const &u16) -> *const &u8 {
    debug x => _1;
    let mut _0: *const &u8;

    bb0: {
        _0 = generic_cast::<&u16, &u8>(copy _1) -> [return: bb1, unwind continue];
    }

    bb1: {
        return;
    }
}
```

I'll break the rest into multiple comments as they are a bit long 😅

ShoyuVanilla commented:
### Extra `mut` on locals

We have another diffs like the following:

```diff
# tests/mir-opt/pre-codegen/slice_index.slice_index_range.PreCodegen.after.panic-abort.mir
                let mut _6: usize;
            }
            scope 4 (inlined core::slice::index::get_offset_len_noubcheck::<u32>) {
-               let _10: *const u32;
+               let mut _10: *const u32;
                scope 5 {
                    let _11: *const u32;
                    scope 6 {
```

Those locals are created while inlining `fn get_offset_len_{mut_}noubcheck`:
https://github.com/rust-lang/rust/blob/0d162b25edd5bf0dba9a22e83b614f1113e90474/library/core/src/slice/index.rs#L81-L103

Yeah, we have `_ as *const T` casts here, again 😄 

And for these test, we have `-Zmir-opt-level=2` so, the code is passed through `SimplifyLocals`.

I simplified the inlined function a bit as the following and ran `rustc -Zmir-opt-level=2 -Zdump-mir=all` with and without this PR.

```rust
#![crate_type = "lib"]

unsafe fn get_offset_len_noubcheck<T>(ptr: *const [T]) {
    let ptr = ptr as *const T;
}
```

Without this PR: no locals are removed as unused in this step.

```diff
diff --git a/mir_dump/bar.get_offset_len_noubcheck.3-2-017.SimplifyLocals-before-const-prop.before.mir b/mir_dump/bar.get_offset_len_noubcheck.3-2-017.SimplifyLocals-before-const-prop.after.mir
index 5771d69fca0..b917ff9fcf2 100644
--- a/mir_dump/bar.get_offset_len_noubcheck.3-2-017.SimplifyLocals-before-const-prop.before.mir
+++ b/mir_dump/bar.get_offset_len_noubcheck.3-2-017.SimplifyLocals-before-const-prop.after.mir
@@ -1,17 +1,17 @@
-// MIR for `get_offset_len_noubcheck` before SimplifyLocals-before-const-prop
+// MIR for `get_offset_len_noubcheck` after SimplifyLocals-before-const-prop
 
 fn get_offset_len_noubcheck(_1: *const [T]) -> () {
     debug ptr => _1;
     let mut _0: ();
     let _2: *const T;
     let mut _3: *const [T];
     scope 1 {
         debug ptr => _2;
     }
 
     bb0: {
         _3 = copy _1;
         _2 = copy _3 as *const T (PtrToPtr);
         return;
     }
 }
```

With this PR: we have extra `let mut _3: *const T;` like in my previous comment. But instead of this, `_2` is unused and truncated. So `let mut _3: *const T;` is shifted by index 1 and becomes `let mut _2: *const T;`, and this is where we get extra `mut`

```diff
diff --git a/mir_dump/bar.get_offset_len_noubcheck.3-2-017.SimplifyLocals-before-const-prop.before.mir b/mir_dump/bar.get_offset_len_noubcheck.3-2-017.SimplifyLocals-before-const-prop.after.mir
index e1e31f66666..db943bb3a18 100644
--- a/mir_dump/bar.get_offset_len_noubcheck.3-2-017.SimplifyLocals-before-const-prop.before.mir
+++ b/mir_dump/bar.get_offset_len_noubcheck.3-2-017.SimplifyLocals-before-const-prop.after.mir
@@ -1,18 +1,17 @@
-// MIR for `get_offset_len_noubcheck` before SimplifyLocals-before-const-prop
+// MIR for `get_offset_len_noubcheck` after SimplifyLocals-before-const-prop
 
 fn get_offset_len_noubcheck(_1: *const [T]) -> () {
     debug ptr => _1;
     let mut _0: ();
-    let _2: *const T;
-    let mut _3: *const T;
-    let mut _4: *const [T];
+    let mut _2: *const T;
+    let mut _3: *const [T];
     scope 1 {
-        debug ptr => _3;
+        debug ptr => _2;
     }
 
     bb0: {
-        _4 = copy _1;
-        _3 = copy _4 as *const T (PtrToPtr);
+        _3 = copy _1;
+        _2 = copy _3 as *const T (PtrToPtr);
         return;
     }
 }
```

@jackh726 would this changes be okay?

lcnr commented:
@rfcbot ping

lcnr commented:
Hi @rust-lang/rust-for-linux, apparently you've started to rely on a type system unsoundness, which should hopefully be easy to fix. Can you point us to where to fix that one or alternatively, do so yourself?

This is quite unfortunate and sorry for the inconvenience :sweat_smile:

BennoLossin commented:
I'm preparing a commit with the diff from above

ojeda committed:
CI: rfl: add temporary commit for compiler bug

The compiler is fixing a bug [1] that Rust for Linux happened to trigger,
thus temporarily add Benno's patch to the CI job.

As usual, the patch will eventually make it to the Linux kernel so that
both sides are good.

Cc: Benno Lossin <lossin@kernel.org>
Link: https://github.com/rust-lang/rust/pull/149389 [1]
Signed-off-by: Miguel Ojeda <ojeda@kernel.org>

rust-timer commented:
Finished benchmarking commit (29fa07e3dbca35d58c0770042f04ba464eb3db4c): [comparison URL](https://perf.rust-lang.org/compare.html?start=f403dee04e6a801fe0f49a9c351fcd2ae8edd37c&end=29fa07e3dbca35d58c0770042f04ba464eb3db4c&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=f403dee04e6a801fe0f49a9c351fcd2ae8edd37c&end=29fa07e3dbca35d58c0770042f04ba464eb3db4c&stat=instructions:u)
Our most reliable metric. Used to determine the overall result above. However, even this metric can be noisy.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.7% | [-0.7%, -0.7%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.7% | [-0.7%, -0.7%] | 1     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=f403dee04e6a801fe0f49a9c351fcd2ae8edd37c&end=29fa07e3dbca35d58c0770042f04ba464eb3db4c&stat=max-rss)
<details>
<summary>Results (primary 1.8%, secondary 2.3%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 5.0%  | [2.9%, 8.8%]   | 3     |
| Regressions ❌ <br /> (secondary)  | 5.9%  | [5.6%, 6.1%]   | 2     |
| Improvements ✅ <br /> (primary)   | -3.0% | [-3.7%, -2.3%] | 2     |
| Improvements ✅ <br /> (secondary) | -1.3% | [-1.6%, -1.0%] | 2     |
| All ❌✅ (primary)                 | 1.8%  | [-3.7%, 8.8%]  | 5     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=f403dee04e6a801fe0f49a9c351fcd2ae8edd37c&end=29fa07e3dbca35d58c0770042f04ba464eb3db4c&stat=cycles:u)
<details>
<summary>Results (secondary 3.4%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 3.4% | [3.4%, 3.4%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=f403dee04e6a801fe0f49a9c351fcd2ae8edd37c&end=29fa07e3dbca35d58c0770042f04ba464eb3db4c&stat=size:linked_artifact)
<details>
<summary>Results (primary 0.0%, secondary 0.0%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.0% | [0.0%, 0.0%] | 9     |
| Regressions ❌ <br /> (secondary)  | 0.0% | [0.0%, 0.0%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.0% | [0.0%, 0.0%] | 9     |
</details>

**Bootstrap**: 481.836s -> 480.714s (-0.23%)
**Artifact size**: 398.12 MiB -> 398.10 MiB (-0.00%)
